### PR TITLE
UnitTest for JwtTokenService

### PR DESCRIPTION
**Pull request by berda97**
Creating UnitTest for JwtTokenSevice

### CODE REVIEW

File: AccountingService/Interface/ITimeService.cs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+﻿namespace AccountingService.Interface
+{
+    public interface ITimeService
+    {
+        int Minutes { get; }
```
Comment (ghost) [inline AccountingService/Interface/ITimeService.cs]: This is not needed here. You can move it to configuration alongside with audience, issuer and secret.

File: AccountingService/Services/JwtTokenService.cs [hunk 2]
```diff
@@ -6,25 +6,31 @@
...
             var authSigningKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_configuration["JwtConfig:Key"]));
-            var expiryInMinutes = Convert.ToInt32(15);
+            var expiryInMinutes = _timeService.Minutes;
+
             var token = new JwtSecurityToken(
                 claims: claims,
                issuer: _configuration["JwtConfig:Issuer"],
                 audience: _configuration["JwtConfig:Audience"],
-                expires: DateTime.Now.AddMinutes(expiryInMinutes),
+                 expires: DateTime.Now.AddMinutes(expiryInMinutes),
```
Comment (ghost) [inline AccountingService/Services/JwtTokenService.cs]: You exposed DateTime.Now from the new TimeService but you are not using it here. Why?

File: UnitTestProject/JwtTokenServiceTest.cs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            Assert.NotNull(token);
+
+            Assert.NotNull(token.Value);
+
+            var tokenHandler = new JwtSecurityTokenHandler();
+            var securityToken = tokenHandler.ReadJwtToken(token.Value);
+            Assert.That(securityToken.Issuer, Is.EqualTo("your_issuer"));
+            Assert.That(securityToken.Audiences.FirstOrDefault(), Is.EqualTo("your_audience"));
+            Assert.IsTrue(securityToken.Claims.Any(c => c.Type == ClaimTypes.Email && c.Value == "Miki123"));
+            Assert.True(securityToken.ValidTo > DateTime.UtcNow);
```
Comment (ghost) [inline UnitTestProject/JwtTokenServiceTest.cs]: Once you adjust your implementation to use the time service, you can moq the value that the interface is returining. And you can then adjust the unit test here as well.

File: UnitTestProject/JwtTokenServiceTest.cs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            var jwtTokenService = new JwtTokenService(_configuration, _timeService);
+            var token = jwtTokenService.GetToken(_claims);
+
+            Assert.NotNull(token);
+            Assert.NotNull(token.Value);
+
+            var tokenHandler = new JwtSecurityTokenHandler();
+            var securityToken = tokenHandler.ReadJwtToken(token.Value);
+
+            // Očekujmo da Audience nije ispravan
```
Comment (ghost) [inline UnitTestProject/JwtTokenServiceTest.cs]: English, please

File: UnitTestProject/JwtTokenServiceTest.cs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            var token = jwtTokenService.GetToken(_claims);
+
+            fakeClock.Advance(Duration.FromMinutes(20));
+
+            Assert.NotNull(token);
+            Assert.NotNull(token.Value);
+
+            var tokenHandler = new JwtSecurityTokenHandler();
+
+            // Očekujte da se baci SecurityTokenExpiredException zbog isteklog vremena
```
Comment (ghost) [inline UnitTestProject/JwtTokenServiceTest.cs]: English, please

File: UnitTestProject/Usings.cs [hunk 6]
```diff
@@ -0,0 +1 @@
+global using NUnit.Framework;
```
Comment (ghost) [inline UnitTestProject/Usings.cs]: Why do you need this?

File: UnitTestProject/UnitTestProject.csproj [hunk 7]
```diff
@@ -0,0 +1,28 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (ghost) [inline UnitTestProject/UnitTestProject.csproj]: The name of the project is usually `<NameOfOriginalProject>.Tests`. In your case `AccountingService.Tests`

File: AccountingService/Services/TimeService.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+﻿using AccountingService.Interface;
+
+namespace AccountingService.Services
+{
+    public class TimeService : ITimeService
+    {
+        public int Minutes { get; }
```
Comment (ghost) [inline AccountingService/Services/TimeService.cs]: This is not necessary now.

File: AccountingService/Services/TimeService.cs [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+        public DateTime Now => DateTime.UtcNow;
+        public TimeService(int minutes)
```
Comment (ghost) [inline AccountingService/Services/TimeService.cs]: Also, this is not needed

File: AccountingService/appsettings.json [hunk 10]
```diff
@@ -8,10 +8,11 @@
   "JwtConfig": {
     "Key": "IiwibmFtZSI6IkpvaG4gRG94cThIIoDvwdueQB468K5xDc5633seEF",
     "Issuer": "https://api.accounting.com",
-    "Audience": "https://api.accounting.com"
+    "Audience": "https://api.accounting.com",
+    "ExpiryInMinutes": 15
   },
   "AppSettings": {
-    "ApiKey": "[Api-Key]"
+    "ApiKey": "[API_KEY]"
```
Comment (ghost) [inline AccountingService/appsettings.json:15]: Why did you change this?

File: UnitTestProject/JwtTokenServiceTest.cs [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+        {
+            var configurationMock = new Mock<IConfiguration>();
+            var timeServiceMock = new Mock<ITimeService>();
+
+            configurationMock.SetupGet(x => x["JwtConfig:Key"]).Returns("your_newly_generated_128_bit_key_here");
+            configurationMock.SetupGet(x => x["JwtConfig:Issuer"]).Returns("your_issuer");
+            configurationMock.SetupGet(x => x["JwtConfig:Audience"]).Returns("your_audience");
+            configurationMock.Setup(x => x["JwtConfig:ExpiryInMinutes"]).Returns("15");
+
+            timeServiceMock.Setup(ts => ts.Now).Returns(() => DateTime.UtcNow.AddMinutes(20));
```
Comment (ghost) [inline UnitTestProject/JwtTokenServiceTest.cs]: Put here a concrete value of DateTime.

File: UnitTestProject/JwtTokenServiceTest.cs (hunk 3 above)
Comment (ghost) [reply UnitTestProject/JwtTokenServiceTest.cs]: Validate here the date that you defined above.

File: AccountingService/Interface/ITimeService.cs (hunk 1 above)
Comment (berda97) [reply AccountingService/Interface/ITimeService.cs]: Done

File: AccountingService/Services/JwtTokenService.cs (hunk 2 above)
Comment (berda97) [reply AccountingService/Services/JwtTokenService.cs]: Done

File: UnitTestProject/JwtTokenServiceTest.cs (hunk 3 above)
Comment (berda97) [reply UnitTestProject/JwtTokenServiceTest.cs]: Done

File: UnitTestProject/JwtTokenServiceTest.cs (hunk 4 above)
Comment (berda97) [reply UnitTestProject/JwtTokenServiceTest.cs]: Done

File: UnitTestProject/JwtTokenServiceTest.cs (hunk 5 above)
Comment (berda97) [reply UnitTestProject/JwtTokenServiceTest.cs]: Done

File: UnitTestProject/Usings.cs (hunk 6 above)
Comment (berda97) [reply UnitTestProject/Usings.cs]: Removed

File: AccountingService.Tests/ExchangeRateServiceTests.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+            var result = exchangeRateService.GetCurrencyExchangeRate(nonNullCurrency);
+            Assert.AreNotEqual(2, result);
+        }
+        [Test]
+        public void GetCurrencyExchangeRate_ReturnsDefaultRateForNonexistentCurrency()
+        {
+            var exchangeRateService = new ExchangeRateService(_configuration);
+            var nonexistentCurrency = "XYZ"; // Zamena sa stvarnom nepostojećom valutom
+
+            // Ako valuta ne postoji ili podaci o kursu nisu ažurirani, očekujemo da se vrati podrazumevana vrednost (1)
```
Comment (ghost) [inline AccountingService.Tests/ExchangeRateServiceTests.cs]: English, please

File: AccountingService/Services/ExchangeRateService.cs [hunk 13]
```diff
@@ -7,28 +7,32 @@ public class ExchangeRateService
     {
         private static JToken _exchangeRates;
         private static DateTime _lastUpdate;
-        private static readonly TimeSpan UpdateInterval = TimeSpan.FromDays(1);
+        public virtual TimeSpan UpdateInterval => TimeSpan.FromDays(1);
```
Comment (ghost) [inline AccountingService/Services/ExchangeRateService.cs]: Why did this change?

File: AccountingService/Services/ExchangeRateService.cs [hunk 14]
```diff
@@ -7,28 +7,32 @@ public class ExchangeRateService
...
-            if (currency == null) 
+            if (currency == null || _exchangeRates == null || _exchangeRates.SelectToken(currency) == null) 
             {
                 return 1;
             }
             ReloadExchangeRates();
             return _exchangeRates[currency].Value<double>();
         }
-        private  void ReloadExchangeRates()
+        public   void ReloadExchangeRates()
```
Comment (ghost) [inline AccountingService/Services/ExchangeRateService.cs]: Why is this public?

**Review by ghost — APPROVED** [review]
LGTM
